<h1>Unit Tests for refine.py</h1>
<p> This notebook is used explain the tests in test_refine.py </p>

<p><b> To run tests use the command line</b>, from the project root run <code>python3 test.py</code></p>

In [1]:
import pandas as pd

In [2]:
test_data = pd.read_csv("data/test_data_refine.csv")

In [3]:
result_nan = pd.read_csv("out/result_test_drop_nan.csv")

In [4]:
result_lang = pd.read_csv("out/result_test_lang_nan_replace.csv")

In [5]:
result_username = pd.read_csv("out/result_test_username_legal.csv")

In [6]:
result_length = pd.read_csv("out/result_test_length_of_tweet.csv")

<p> This test file has a number of invalid data values in it which will be removed by running the methods in <i>refine.py</i>. To test that the functions were removing the correct data <i>test_refine.py</i> was written</p>

In [7]:
test_data


id_str                                          from_user  \
0    540930409279008000                                   text@!...TooLong   
1  5.40929229052203e+17                                                NaN   
2  5.40926888768966e+17                                   inv@lidU$er-name   
3  5.40918061042515e+17  thisUserNameIsClearlyTooLongToBeValidAndTheref...   
4       clearly invalid                                                NaN   
5  5.40926211594416e+17                                     duplicateTweet   
6  5.40926211594416e+17                                     duplicateTweet   
7  5.40923325061808e+17                                           nullLang   
8  5.40922477552369e+17                                        capitalLang   

                                                text  \
0  Here's a message that is longer than 140 chars...   
1  RT @shaunmajumder: Feels good to be the @CBCDa...   
2  RT @shaunmajumder: Feels good to be the @CBCDa...   
3  With @NASA_Orion today &amp; @Philae2014 @ESA_...   
4                                                NaN   
5  RT @VersaTechnology: Congratulations @Philae20...   
6  RT @VersaTechnology: Congratulations @Philae20...   
7  RT @costumepups: #CometLanding http://t.co/7Rj...   
8  RT @glxp: From the #CometLanding to the Moon--...   

                       created_at              time  geo_coordinates  \
0  Fri Dec 05 18:07:14 +0000 2014  05/12/2014 18:07              NaN   
1  Fri Dec 05 18:02:32 +0000 2014  05/12/2014 18:02              NaN   
2  Fri Dec 05 17:53:14 +0000 2014  05/12/2014 17:53              NaN   
3  Fri Dec 05 17:18:10 +0000 2014  05/12/2014 17:18              NaN   
4                             NaN               NaN              NaN   
5  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
6  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
7  Fri Dec 05 17:39:05 +0000 2014  05/12/2014 17:39              NaN   
8  Fri Dec 05 17:35:43 +0000 2014  05/12/2014 17:35              NaN   

  user_lang  in_reply_to_user_id_str  in_reply_to_screen_name  \
0        en                      NaN                      NaN   
1        en                      NaN                      NaN   
2        en                      NaN                      NaN   
3        en                      NaN                      NaN   
4       NaN                      NaN                      NaN   
5        en                      NaN                      NaN   
6        en                      NaN                      NaN   
7       NaN                      NaN                      NaN   
8        EN                      NaN                      NaN   

   from_user_id_str  in_reply_to_status_id_str  \
0      1.297570e+09                        NaN   
1      1.772769e+08                        NaN   
2      2.461882e+07                        NaN   
3      1.120676e+08                        NaN   
4               NaN                        NaN   
5      1.386198e+09                        NaN   
6      1.386198e+09                        NaN   
7      1.494499e+09                        NaN   
8      7.645674e+08                        NaN   

                                              source  \
0  <a href="http://twitter.com" rel="nofollow">Tw...   
1  <a href="http://twitter.com/download/android" ...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://bufferapp.com" rel="nofollow">...   
4                                                NaN   
5  <a href="http://twitter.com" rel="nofollow">Tw...   
6  <a href="http://twitter.com" rel="nofollow">Tw...   
7  <a href="http://twitter.com/download/iphone" r...   
8  <a href="http://twitter.com" rel="nofollow">Tw...   

                                   profile_image_url  user_followers_count  \
0  http://pbs.twimg.com/profile_images/3555068293...                 425.0   
1  http://pbs.twimg.com/profile_images/4590694303...                4331.0   
2  http

<h2> Test: test_length_of_tweet</h2>
<p> The function this tests is <code>textCheck()</code>, which seeks to remove a row which contains a tweet that is longer than 140 characters. This could arise through data corruption elsewhere and is important to check. In the example data above row 0 is removed, along with 1, 4 and 7 - by <code>removeNAN()</code> as described below.</p>

<p>Retweets get an extra 20 characters as there is extra text appended to the normal tweet length to indicate the user who originally tweeted it: RT@"up to 15 chars for username"</p>

In [8]:
def test_length_of_tweet(self):
        df = rf.removeNAN(self.df)
        length = len(df)
        df = rf.textCheck(df)
        self.assertFalse(length == len(df)) # check that rows have been removed
        df.to_csv("tests/out/result_test_length_of_tweet.csv");

<h3> Output: </h3>
<p> The output for this function confirms the hypothesis above. Row 0 was removed, along with rows 1,4,7.</p>

In [9]:
result_length

Unnamed: 0        id_str  \
0           2  5.409269e+17   
1           3  5.409181e+17   
2           5  5.409262e+17   
3           6  5.409262e+17   
4           8  5.409225e+17   

                                           from_user  \
0                                   inv@lidU$er-name   
1  thisUserNameIsClearlyTooLongToBeValidAndTheref...   
2                                     duplicateTweet   
3                                     duplicateTweet   
4                                        capitalLang   

                                                text  \
0  RT @shaunmajumder: Feels good to be the @CBCDa...   
1  With @NASA_Orion today &amp; @Philae2014 @ESA_...   
2  RT @VersaTechnology: Congratulations @Philae20...   
3  RT @VersaTechnology: Congratulations @Philae20...   
4  RT @glxp: From the #CometLanding to the Moon--...   

                       created_at              time  geo_coordinates  \
0  Fri Dec 05 17:53:14 +0000 2014  05/12/2014 17:53              NaN   
1  Fri Dec 05 17:18:10 +0000 2014  05/12/2014 17:18              NaN   
2  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
3  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
4  Fri Dec 05 17:35:43 +0000 2014  05/12/2014 17:35              NaN   

  user_lang  in_reply_to_user_id_str  in_reply_to_screen_name  \
0        en                      NaN                      NaN   
1        en                      NaN                      NaN   
2        en                      NaN                      NaN   
3        en                      NaN                      NaN   
4        EN                      NaN                      NaN   

   from_user_id_str  in_reply_to_status_id_str  \
0      2.461882e+07                        NaN   
1      1.120676e+08                        NaN   
2      1.386198e+09                        NaN   
3      1.386198e+09                        NaN   
4      7.645674e+08                        NaN   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://bufferapp.com" rel="nofollow">...   
2  <a href="http://twitter.com" rel="nofollow">Tw...   
3  <a href="http://twitter.com" rel="nofollow">Tw...   
4  <a href="http://twitter.com" rel="nofollow">Tw...   

                                   profile_image_url  user_followers_count  \
0  http://pbs.twimg.com/profile_images/4204092788...               41655.0   
1  http://pbs.twimg.com/profile_images/4866582907...                2010.0   
2  http://pbs.twimg.com/profile_images/3611682679...                 214.0   
3  http://pbs.twimg.com/profile_images/3611682679...                 214.0   
4  http://pbs.twimg.com/profile_images/2590257372...                4676.0   

   user_friends_count                                         status_url  \
0               745.0  http://twitter.com/markcritch/statuses/5409268...   
1              2210.0  http://twitter.com/ChathamPhoto/statuses/54091...   
2               106.0  http://twitter.com/vencivelev/statuses/5409262...   
3               106.0  http://twitter.com/vencivelev/statuses/5409262...   
4              5056.0  http://twitter.com/TerryMcCracken/statuses/540...   

                                        entities_str  
0  {"hashtags":[{"text":"MiniMansbridge","indices...  
1  {"hashtags":[{"text":"STEM","indices":[99,104]...  
2  {"hashtags":[{"text":"Philae","indices":[49,56...  
3  {"hashtags":[{"text":"Philae","indices":[49,56...  
4  {"hashtags":[{"text":"CometLanding","indices":...

<h2> Test: test_username_legal</h2>
<p> The function this tests is <code>fromUserCheck()</code>, which seeks to remove a row that contains a username which does not follow the twitter rules. This could arise through data corruption elsewhere and is important to check. In the example data above rows 5,6 and 8 will stay.</p>

In [10]:
def test_username_legal(self):
        length = len(self.df)
        df = rf.removeNAN(self.df)
        df = rf.fromUserCheck(df)
        self.assertFalse(length == len(df)) # check that rows have been removed
        df.to_csv("tests/out/result_test_username_legal.csv");

<h3> Output: </h3>
<p> The output for this function confirms the hypothesis above. The user_lang value in rows 5, 6 and 8 was replaced.</p>

In [11]:
result_username


Unnamed: 0        id_str       from_user  \
0           5  5.409262e+17  duplicateTweet   
1           6  5.409262e+17  duplicateTweet   
2           8  5.409225e+17     capitalLang   

                                                text  \
0  RT @VersaTechnology: Congratulations @Philae20...   
1  RT @VersaTechnology: Congratulations @Philae20...   
2  RT @glxp: From the #CometLanding to the Moon--...   

                       created_at              time  geo_coordinates  \
0  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
1  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
2  Fri Dec 05 17:35:43 +0000 2014  05/12/2014 17:35              NaN   

  user_lang  in_reply_to_user_id_str  in_reply_to_screen_name  \
0        en                      NaN                      NaN   
1        en                      NaN                      NaN   
2        EN                      NaN                      NaN   

   from_user_id_str  in_reply_to_status_id_str  \
0      1.386198e+09                        NaN   
1      1.386198e+09                        NaN   
2      7.645674e+08                        NaN   

                                              source  \
0  <a href="http://twitter.com" rel="nofollow">Tw...   
1  <a href="http://twitter.com" rel="nofollow">Tw...   
2  <a href="http://twitter.com" rel="nofollow">Tw...   

                                   profile_image_url  user_followers_count  \
0  http://pbs.twimg.com/profile_images/3611682679...                 214.0   
1  http://pbs.twimg.com/profile_images/3611682679...                 214.0   
2  http://pbs.twimg.com/profile_images/2590257372...                4676.0   

   user_friends_count                                         status_url  \
0               106.0  http://twitter.com/vencivelev/statuses/5409262...   
1               106.0  http://twitter.com/vencivelev/statuses/5409262...   
2              5056.0  http://twitter.com/TerryMcCracken/statuses/540...   

                                        entities_str  
0  {"hashtags":[{"text":"Philae","indices":[49,56...  
1  {"hashtags":[{"text":"Philae","indices":[49,56...  
2  {"hashtags":[{"text":"CometLanding","indices":...

<h2> Test: test_lang_nan_replace</h2>
<p> The function this tests is <code>langCheck()</code>, which seeks to replace a null value in the user_lang column with default value "en". This is because there may be a reason why the langauge cannot be detected but the tweet itself is not corrupted. English was chosen as it is the most common language setting among twitter users. The function also converts everything to lowercase. In the example data above, row 7 user_lang value will change from NaN to en. </p>

In [12]:
def test_lang_nan_replace(self):
        #a null value present
        self.assertTrue(self.df["user_lang"].isnull().any())
        df = rf.langCheck(self.df)
        # nulls replaced with en
        self.assertFalse(df["user_lang"].isnull().any())
        df.to_csv("tests/out/result_test_lang_nan_replace.csv");

<h3> Output: </h3>
<p> The output for this function confirms the hypothesis above. The user_lang value in row 7 was replaced.</p>

In [13]:
result_lang

Unnamed: 0                id_str  \
0           0    540930409279008000   
1           1  5.40929229052203e+17   
2           2  5.40926888768966e+17   
3           3  5.40918061042515e+17   
4           4       clearly invalid   
5           5  5.40926211594416e+17   
6           6  5.40926211594416e+17   
7           7  5.40923325061808e+17   
8           8  5.40922477552369e+17   

                                           from_user  \
0                                   text@!...TooLong   
1                                                NaN   
2                                   inv@lidU$er-name   
3  thisUserNameIsClearlyTooLongToBeValidAndTheref...   
4                                                NaN   
5                                     duplicateTweet   
6                                     duplicateTweet   
7                                           nullLang   
8                                        capitalLang   

                                                text  \
0  Here's a message that is longer than 140 chars...   
1  RT @shaunmajumder: Feels good to be the @CBCDa...   
2  RT @shaunmajumder: Feels good to be the @CBCDa...   
3  With @NASA_Orion today &amp; @Philae2014 @ESA_...   
4                                                NaN   
5  RT @VersaTechnology: Congratulations @Philae20...   
6  RT @VersaTechnology: Congratulations @Philae20...   
7  RT @costumepups: #CometLanding http://t.co/7Rj...   
8  RT @glxp: From the #CometLanding to the Moon--...   

                       created_at              time  geo_coordinates  \
0  Fri Dec 05 18:07:14 +0000 2014  05/12/2014 18:07              NaN   
1  Fri Dec 05 18:02:32 +0000 2014  05/12/2014 18:02              NaN   
2  Fri Dec 05 17:53:14 +0000 2014  05/12/2014 17:53              NaN   
3  Fri Dec 05 17:18:10 +0000 2014  05/12/2014 17:18              NaN   
4                             NaN               NaN              NaN   
5  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
6  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
7  Fri Dec 05 17:39:05 +0000 2014  05/12/2014 17:39              NaN   
8  Fri Dec 05 17:35:43 +0000 2014  05/12/2014 17:35              NaN   

  user_lang  in_reply_to_user_id_str  in_reply_to_screen_name  \
0        en                      NaN                      NaN   
1        en                      NaN                      NaN   
2        en                      NaN                      NaN   
3        en                      NaN                      NaN   
4        en                      NaN                      NaN   
5        en                      NaN                      NaN   
6        en                      NaN                      NaN   
7        en                      NaN                      NaN   
8        en                      NaN                      NaN   

   from_user_id_str  in_reply_to_status_id_str  \
0      1.297570e+09                        NaN   
1      1.772769e+08                        NaN   
2      2.461882e+07                        NaN   
3      1.120676e+08                        NaN   
4               NaN                        NaN   
5      1.386198e+09                        NaN   
6      1.386198e+09                        NaN   
7      1.494499e+09                        NaN   
8      7.645674e+08                        NaN   

                                              source  \
0  <a href="http://twitter.com" rel="nofollow">Tw...   
1  <a href="http://twitter.com/download/android" ...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://bufferapp.com" rel="nofollow">...   
4                                                NaN   
5  <a href="http://twitter.com" rel="nofollow">Tw...   
6  <a href="http://twitter.com" rel="nofollow">Tw...   
7  <a href="http://twitter.com/download/iphone" r...   
8  <a href="http://twitter.com" rel="nofollow">Tw...   

                                   profile_image_url  us

<h2> Test: test_drop_nan</h2>
<p> The function this tests is <i>removeNAN()</i> which seeks to remove a row that has null values, where these null values belong to a column with a minority of null values. In the example data above, row 4 will be removed, along with row 1 and row 7</p>

In [14]:
def test_drop_nan(self):
        first_len = len(self.df)
        df = rf.removeNAN(self.df)
        self.assertFalse(first_len == len(df)) # check that rows have been removed
        df.to_csv("tests/out/result_test_drop_nan.csv");

<h3> Output: </h3>
<p> The output for this function confirms the hypothesis above. 3 rows were removed, identifiable by their id_str</p>

In [15]:
result_nan

Unnamed: 0        id_str  \
0           0  5.409304e+17   
1           2  5.409269e+17   
2           3  5.409181e+17   
3           5  5.409262e+17   
4           6  5.409262e+17   
5           8  5.409225e+17   

                                           from_user  \
0                                   text@!...TooLong   
1                                   inv@lidU$er-name   
2  thisUserNameIsClearlyTooLongToBeValidAndTheref...   
3                                     duplicateTweet   
4                                     duplicateTweet   
5                                        capitalLang   

                                                text  \
0  Here's a message that is longer than 140 chars...   
1  RT @shaunmajumder: Feels good to be the @CBCDa...   
2  With @NASA_Orion today &amp; @Philae2014 @ESA_...   
3  RT @VersaTechnology: Congratulations @Philae20...   
4  RT @VersaTechnology: Congratulations @Philae20...   
5  RT @glxp: From the #CometLanding to the Moon--...   

                       created_at              time  geo_coordinates  \
0  Fri Dec 05 18:07:14 +0000 2014  05/12/2014 18:07              NaN   
1  Fri Dec 05 17:53:14 +0000 2014  05/12/2014 17:53              NaN   
2  Fri Dec 05 17:18:10 +0000 2014  05/12/2014 17:18              NaN   
3  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
4  Fri Dec 05 17:50:33 +0000 2014  05/12/2014 17:50              NaN   
5  Fri Dec 05 17:35:43 +0000 2014  05/12/2014 17:35              NaN   

  user_lang  in_reply_to_user_id_str  in_reply_to_screen_name  \
0        en                      NaN                      NaN   
1        en                      NaN                      NaN   
2        en                      NaN                      NaN   
3        en                      NaN                      NaN   
4        en                      NaN                      NaN   
5        EN                      NaN                      NaN   

   from_user_id_str  in_reply_to_status_id_str  \
0      1.297570e+09                        NaN   
1      2.461882e+07                        NaN   
2      1.120676e+08                        NaN   
3      1.386198e+09                        NaN   
4      1.386198e+09                        NaN   
5      7.645674e+08                        NaN   

                                              source  \
0  <a href="http://twitter.com" rel="nofollow">Tw...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://bufferapp.com" rel="nofollow">...   
3  <a href="http://twitter.com" rel="nofollow">Tw...   
4  <a href="http://twitter.com" rel="nofollow">Tw...   
5  <a href="http://twitter.com" rel="nofollow">Tw...   

                                   profile_image_url  user_followers_count  \
0  http://pbs.twimg.com/profile_images/3555068293...                 425.0   
1  http://pbs.twimg.com/profile_images/4204092788...               41655.0   
2  http://pbs.twimg.com/profile_images/4866582907...                2010.0   
3  http://pbs.twimg.com/profile_images/3611682679...                 214.0   
4  http://pbs.twimg.com/profile_images/3611682679...                 214.0   
5  http://pbs.twimg.com/profile_images/2590257372...                4676.0   

   user_friends_count                                         status_url  \
0               113.0  http://twitter.com/amika0078788556/statuses/54...   
1               745.0  http://twitter.com/markcritch/statuses/5409268...   
2              2210.0  http://twitter.com/ChathamPhoto/statuses/54091...   
3               106.0  http://twitter.com/vencivelev/statuses/5409262...   
4               106.0  http://twitter.com/vencivelev/statuses/5409262...   
5              5056.0  http://twitter.com/TerryMcCracken/statuses/540...   

                                        entities_str  
0  {"hashtags":[{"text":"Philae","indices":[49,56...  
1  {"hashtags":[{"text":"MiniMansbridge","indices...  
2  {"hashtags":[{"text":"STEM","indices":[99,104]